### ISSUE TITLE
Revise selecting best candidate policy

### ISSUE THREAD

**Issue by junhoseo** [WICG/spatial-navigation#94]
Currently we consider distance and order in DOM structure. But sometimes these are not sufficient for reasonable choice.

Case 1:
![image](https://user-images.githubusercontent.com/4897376/44316189-4faa1980-a465-11e8-8e36-2b9077a52c29.png)

In above case, distance with active element is same each other. So blue element will be selected if it is earlier element in DOM tree, and yellow element will be selected on the contrary.

Case 2:
![image](https://user-images.githubusercontent.com/4897376/44316208-66507080-a465-11e8-9074-3a6b07fd4065.png)

![image](https://user-images.githubusercontent.com/4897376/44316263-add6fc80-a465-11e8-8437-687243208f80.png)

Sometimes a focusable element can be placed inside another focusable element. In this case, should we allow focus moving to opposite element? Opposite element is not placed in direction, but partially.

**Comment by jihyerish** [comment]
>  Case 1.

Yes. The focus will move depending on the DOM order if the candidates have the same distance from the active element.
So far, I think it's a reasonable result. But in some condition, it would be nice to consider the visual inclusion relation among the elements.

> Case 2.

In that regard, we need to consider the visual inclusion relation in this case.
If you see the sample below,
https://wicg.github.io/spatial-navigation/demo/sample/heuristic_text_type_elements.html
![image](https://user-images.githubusercontent.com/6636090/44983111-0c968d00-afb3-11e8-8a15-525c91f3c349.png)

The content area of the (A)`anchor element` with "spatial navigation" contains that of the (B)`anchor element` with "repository".
This causes that the focus cannot move from (A) to (B) nor from (B) to (A).

**Comment by junhoseo** [comment]
For case #2, and minimizing unreachable case, I suggest an additional step 'internal search' in processing model. Let's see following scenario.

![image](https://user-images.githubusercontent.com/4897376/46132765-eabebc00-c279-11e8-9305-75011b64d27a.png)

In this scenario, currently red color element is active element. 3 elements are children of a container element(blue color). Meanwhile, there is green element and it is (visually) inside active element. And let it is not child of the container. Usually, such green element never gets focus because it is encapsulated(In this scenario, green element can get focus through yellow element. But it is introduced to show an important case.). And we hit down key.


For now, let's see how 'internal search' step works.

![image](https://user-images.githubusercontent.com/4897376/46133013-77697a00-c27a-11e8-9cb1-bd080dc21839.png)

In internal search step, valid rectangle is defined. It is rectangle of active element, and limits focus candidate. Also we define starting point as edge of active element. In this scenario, only green element is candidate for getting focus in internal search step.

Important different point between internal search for active element and finding in container is, in case of internal search, candidate should be 'fully' inside of valid rect. But, in case of finding in container, partially visible element can be candidate. This is why yellow element can't be candidate.


I think we can integrate this step with existing processing model.

```
For each container from active element's nearest container to root container
    if internal search step finds a candidate
        stop loop

    if normal step finds a candidate or scroll
        stop loop
```


So, let's see, how integrated processing model works.
In first iteration, container is blue element. In this iteration, we can't any candidate(or scroll) in neither internal search step nor normal step.
In second iteration, container is root element. In this iteration, green element can be candidate in internal search step. So we give focus to the candidate and stop loop.

**Comment by hugoholgersson** [comment]
> Usually, such green element never gets focus because it is encapsulated

The problem is that spatnav only looks _outside_ of activeElement's visual **box**, right?

> Also we define starting point as edge of active element.

Perhaps this is all we need to do (no need for an extra "internal" step)?

I wonder if we ever need to use the complete "focus **box**".... Perhaps [edges](https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/page/spatial_navigation.cc?q=function:OppositeEdge&sq=package:chromium) as search origin give good enough approximations for most (all?) distance calculations?

**Comment by hugoholgersson** [comment]
> For example, in above scenario, left(apricot? color) and right(yellow color) can be candidate when we hit down key. It looks like such elements are not placed in direction. Of course, it depends on the person. What do you think?

Yes, they _can_, which is good. All **visible** focusables in the direction, down, should be considered candidates. The green focusable will anyway win because it's closer to activeElement.

**Comment by junhoseo** [comment]
:)
OK. So this is more harder example :) :) :)

![image](https://user-images.githubusercontent.com/4897376/46195771-2a4fdb80-c340-11e8-9b6d-68294ddc9f4e.png)

Focusable elements are placed as grid. And red color element is active element.
We hit down key again.

I think most people expect yellow element will get focus. But actually, green element will get focus because it is placed in direction and closest element with red element. How we can resolve this problem when we use active element's edge as starting point in common cases?

**Comment by hugoholgersson** [comment]
I agree. :) I get your point...

One alternative could be to prioritize elements that are _fully_ aligned (i.e. elements that can be [partly] projected onto each other)...?

**Comment by frivoal** [comment]
I think the suggestion in https://github.com/WICG/spatial-navigation/issues/94#issuecomment-425004503 is actually quite good. Should we write it in the spec first, or first try to implement in the polyfill, try it to see if it is nice, and update the spec only if we like it?

**Comment by frivoal** [comment]
For case 1 in https://github.com/WICG/spatial-navigation/issues/94#issue-351959505, I am much less sure what the order should be if not the DOM. The one most in front? The largest one? The one with the largest **visible** area? I am not 100% sure how to decide which answer is best.

**Comment by frivoal** [comment]
I think there are several issues being discussed at the same time in this one. Separating them into sub topics: 

* case 1 form https://github.com/WICG/spatial-navigation/issues/94#issue-351959505 is now here: https://github.com/WICG/spatial-navigation/issues/121

* Prioritizing fully or mostly aligned elements is now here: https://github.com/WICG/spatial-navigation/issues/122

As for the rest, which I think is best described in https://github.com/WICG/spatial-navigation/issues/94#issuecomment-425004503, I think I was actually confused when I said I agreed. In https://github.com/WICG/spatial-navigation/issues/94#issuecomment-425004503, if the green box is in the same container as the other boxes, the specification already handles that. However, if it is not in the same container, then it seems that skipping it is a feature, not a bug. That is what containers are for. If that's not desired, then making the "blue" box in that example a container seems like a mistake.

Is there any situation where we would want the blue box to be a container, but still go from red to green?

**Comment by junhoseo** [comment]
Hi @frivoal,
> if the green box is in the same container as the other boxes, the specification already handles that.
Would you please explain how spec handles?

> Is there any situation where we would want the blue box to be a container, but still go from red to green?
I found a case:
![image](https://user-images.githubusercontent.com/4897376/47132049-e9af0680-d2dc-11e8-9c83-63d4a751322e.png)

In above case, a link "arrow keys" is placed in two lines. So the layout box of the element will be red box. Meanwhile, another link "modifier key" is placed in red box. So in this case, we hardly navigate into "modifier key.

**Comment by junhoseo** [comment]
> I found a case:

Oh, the case is already noticed by @jihyerish at https://github.com/WICG/spatial-navigation/issues/94#issuecomment-418079301 :)

**Comment by frivoal** [comment]
Issue migrated to w3c/csswg-drafts#3386
